error_writer.go

```diff
@@ -54,69 +52,54 @@ type ErrorWriter struct {
-		writer.unaryConnectContentTypes[unary] = struct{}{}
-		streaming := connectContentTypeFromCodecName(StreamTypeBidi, name)
-		writer.streamingConnectContentTypes[streaming] = struct{}{}
-	}
-	if config.HandleGRPC {
-		writer.grpcContentTypes[grpcContentTypeDefault] = struct{}{}
-		for name := range config.Codecs {
-			ct := grpcContentTypeFromCodecName(false /* web */, name)
-			writer.grpcContentTypes[ct] = struct{}{}
-		}
-	}
-	if config.HandleGRPCWeb {
-		writer.grpcWebContentTypes[grpcWebContentTypeDefault] = struct{}{}
-		for name := range config.Codecs {
-			ct := grpcContentTypeFromCodecName(true /* web */, name)
-			writer.grpcWebContentTypes[ct] = struct{}{}
-		}
-	}
-	return writer
 }
 
 func (w *ErrorWriter) classifyRequest(request *http.Request) protocolType {
 	ctype := canonicalizeContentType(getHeaderCanonical(request.Header, headerContentType))
-	if _, ok := w.unaryConnectContentTypes[ctype]; ok {
-		if err := connectCheckProtocolVersion(request, w.requireConnectProtocolHeader); err != nil {
-			return unknownProtocol
+	method := request.Method
+	switch {
+	case w.handleGRPC && (ctype == grpcContentTypeDefault || strings.HasPrefix(ctype, grpcContentTypePrefix)):
+		if method != http.MethodPost {
+			break
+		}
+		return grpcProtocol
+	case w.handleGRPCWeb && (ctype == grpcWebContentTypeDefault || strings.HasPrefix(ctype, grpcWebContentTypePrefix)):
+		if method != http.MethodPost {
+			break
+		}
+		return grpcWebProtocol
+	case strings.HasPrefix(ctype, connectStreamingContentTypePrefix):
+		if method != http.MethodPost {
+			break
 		}
-		return connectUnaryProtocol
-	}
-	if _, ok := w.streamingConnectContentTypes[ctype]; ok {
 		if err := connectCheckProtocolVersion(request, w.requireConnectProtocolHeader); err != nil {
-			return unknownProtocol
+			break
 		}
 		return connectStreamProtocol
-	}
-	if _, ok := w.grpcContentTypes[ctype]; ok {
-		return grpcProtocol
-	}
-	if _, ok := w.grpcWebContentTypes[ctype]; ok {
-		return grpcWebProtocol
-	}
-	// Check for Connect-Protocol-Version header or connect protocol query
-	// parameter to support connect GET requests.
-	if request.Method == http.MethodGet {
-		connectVersion := getHeaderCanonical(request.Header, connectProtocolVersion)
-		if connectVersion == connectProtocolVersion {
-			return connectUnaryProtocol
+	case strings.HasPrefix(ctype, connectUnaryContentTypePrefix):
+		if method != http.MethodPost {
+			break
 		}
-		connectVersion = request.URL.Query().Get(connectUnaryConnectQueryParameter)
-		if connectVersion == connectUnaryConnectQueryValue {
-			return connectUnaryProtocol
+		if err := connectCheckProtocolVersion(request, w.requireConnectProtocolHeader); err != nil {
+			break
 		}
+		return connectUnaryProtocol
+	case ctype == "":
+		if method != http.MethodGet {
+			break
+		}
+		if err := connectCheckProtocolVersion(request, w.requireConnectProtocolHeader); err != nil {
+			break
+		}
+		return connectUnaryProtocol
```

Comment:
I'm finding the big switch with nested `break`s difficult to read. It'd be clearer to me if we had an `isPost := method == http.MethodPost` before the switch and inlined the HTTP verb checks into the case statements where possible.